src/boot/linux.c

```diff
@@ -119,16 +93,73 @@ EFI_STATUS linux_exec(
         if (err != EFI_SUCCESS)
                 return log_error_status(err, "Bad kernel image: %m");
 
-        _cleanup_(unload_imagep) EFI_HANDLE kernel_image = NULL;
-        err = load_image(parent, kernel->iov_base, kernel->iov_len, &kernel_image);
+        if (secure_boot_enabled() && shim_load_image_is_shim())
+                return load_with_shim(parent, compat_entry_point, kernel, initrd);
+
+        err = pe_kernel_check_no_relocation(kernel->iov_base);
+        if (err != EFI_SUCCESS)
+                return err;
+
+        const PeSectionHeader *headers;
+        size_t n_headers;
+
+        /* Do we need to validate anyting here? the len? */
+        err = pe_section_table_from_base(kernel->iov_base, &headers, &n_headers);
         if (err != EFI_SUCCESS)
-                return log_error_status(err, "Error loading kernel image: %m");
+                return log_error_status(err, "Cannot read sections: %m");
+
+        /* Do we need to ensure under 4gb address on x86? */
+        _cleanup_pages_ Pages loaded_kernel_pages = xmalloc_pages(
+                        AllocateAnyPages, EfiLoaderCode, EFI_SIZE_TO_PAGES(kernel_size_in_memory), 0);
+
+        uint8_t* loaded_kernel = PHYSICAL_ADDRESS_TO_POINTER(loaded_kernel_pages.addr);
+        FOREACH_ARRAY(h, headers, n_headers) {
+                if (h->PointerToRelocations != 0)
+                        return log_error_status(EFI_LOAD_ERROR, "Inner kernel image contains sections with relocations, which we do not support.");
+                if (h->SizeOfRawData == 0)
+                        continue;
+
+                if ((h->VirtualAddress < image_base)
+                    || (h->VirtualAddress - image_base + h->SizeOfRawData > kernel_size_in_memory))
+                        return log_error_status(EFI_LOAD_ERROR, "Section would write outside of memory");
+                memcpy(loaded_kernel + h->VirtualAddress - image_base,
+                       (const uint8_t*)kernel->iov_base + h->PointerToRawData,
+                       h->SizeOfRawData);
+                if (h->VirtualSize > h->SizeOfRawData)
+                        memzero(loaded_kernel + h->VirtualAddress + h->SizeOfRawData,
```

Comment:
drop the conditionalization here btw. memzero() is explicitly fine with a zero length